### PR TITLE
sort by score, fix sort by null & missing range value

### PR DESCRIPTION
**Pull request by StigNorland**
https://unit.atlassian.net/browse/NP-46537 Indexeringsfeil av poster med flere tusen forfattere
https://unit.atlassian.net/browse/NP-46707  Wrong pagination on results search
https://unit.atlassian.net/browse/NP-46120  Diskusjon om "fremhevede" resultat

### CODE REVIEW

File: search-commons/src/main/java/no/unit/nva/search2/common/SearchQuery.java [hunk 1]
```diff
@@ -272,11 +273,19 @@ protected QueryBuilder builderSearchAllQuery(K searchAllKey) {
...
+        }
+        setFetchSource(builder);
+    }
+
     private URI nextResultsBySortKey(SwsResponse response, Map<String, String> requestParameter, URI gatewayUri) {
         requestParameter.remove(Words.FROM);
         var sortParameter =
             response.getSort().stream()
-                .map(Object::toString)
+                .map(value -> nonNull(value) ? value : "null")
```
Comment (StigNorland) [inline search-commons/src/main/java/no/unit/nva/search2/common/SearchQuery.java:288]: Opensearch search_after 'null' bug fix

File: search-commons/src/main/java/no/unit/nva/search2/common/builder/OpensearchQuery.java [hunk 2]
```diff
@@ -24,20 +25,32 @@ public abstract class OpensearchQuery<K extends Enum<K> & ParameterKey> {
 
     public QueryTools<K> queryTools = new QueryTools<>();
 
+    protected abstract Stream<Entry<K, QueryBuilder>> buildMatchAnyKeyValuesQuery(K key, String... values);
+
+    protected abstract Stream<Entry<K, QueryBuilder>> buildMatchAllValuesQuery(K key, String... values);
+
     public Stream<Map.Entry<K, QueryBuilder>> buildQuery(K key, String value) {
-        final var values = value.split(COMMA);
+        final var values = splitAndFixMissingRangeValue(key, value);
```
Comment (StigNorland) [inline search-commons/src/main/java/no/unit/nva/search2/common/builder/OpensearchQuery.java:33]: Missing range value fix

File: search-commons/src/main/java/no/unit/nva/search2/resource/ResourceSearchQuery.java [hunk 3]
```diff
@@ -356,9 +353,6 @@ protected void setValue(String key, String value) {
                 case SORT_ORDER -> mergeToKey(SORT, decodedValue);
                 case PUBLICATION_LANGUAGE, PUBLICATION_LANGUAGE_NOT,
                      PUBLICATION_LANGUAGE_SHOULD -> searchQuery.parameters().set(qpKey, expandLanguage(decodedValue));
-                case CREATED_BEFORE, CREATED_SINCE,
-                     MODIFIED_BEFORE, MODIFIED_SINCE,
-                     PUBLISHED_BEFORE, PUBLISHED_SINCE -> searchQuery.parameters().set(qpKey, expandYearToDate(decodedValue));
```
Comment (StigNorland) [inline search-commons/src/main/java/no/unit/nva/search2/resource/ResourceSearchQuery.java:361]: expansion done in querybuilder

File: search-commons/src/main/java/no/unit/nva/search2/ticket/TicketSearchQuery.java [hunk 4]
```diff
@@ -262,8 +258,6 @@ protected void setValue(String key, String value) {
                 case FIELDS -> searchQuery.parameters().set(qpKey, ignoreInvalidFields(decodedValue));
                 case SORT -> mergeToKey(SORT, trimSpace(decodedValue));
                 case SORT_ORDER -> mergeToKey(SORT, decodedValue);
-                case CREATED_DATE, MODIFIED_DATE, PUBLICATION_MODIFIED_DATE ->
-                    searchQuery.parameters().set(qpKey, expandIfMissingRange(decodedValue));
```
Comment (StigNorland) [inline search-commons/src/main/java/no/unit/nva/search2/ticket/TicketSearchQuery.java:266]: moved to QueryBuilder

File: search-commons/src/test/java/no/unit/nva/search2/ImportCandidateClientTest.java [hunk 5]
```diff
@@ -242,11 +244,22 @@ static Stream<URI> uriProvider() {
...
+                URI.create(REQUEST_BASE_URL + "MODIFIED_DATE=2023-05,&size=7"),
                 URI.create(REQUEST_BASE_URL + "PUBLICATION_YEAR_BEFORE=2023&size=5"),
-                URI.create(REQUEST_BASE_URL + "publication_year=2022,2022&size=1"),
                 URI.create(REQUEST_BASE_URL + "PublicationYearBefore=2024&publication_year_since=2023&size=3"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=2022,2022&size=1"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=2022&size=1"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=,2022&size=5"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=2022,&size=4"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=2023&size=3"),
+                URI.create(REQUEST_BASE_URL + "publicationYear=2022,2023&size=4"),
```
Comment (StigNorland) [inline search-commons/src/test/java/no/unit/nva/search2/ImportCandidateClientTest.java:262]: more range search tests

File: search-commons/src/main/java/no/unit/nva/search2/common/constant/Functions.java [hunk 6]
```diff
@@ -114,9 +113,9 @@ public static String trimSpace(String value) {
         return value.replaceAll(PATTERN_IS_ASC_OR_DESC_GROUP, PATTERN_IS_SELECTED_GROUP);
     }
 
-    public static String expandYearToDate(String value) {
-        return value.length() == 4 ? value + JANUARY_FIRST : value;
-    }
+//    public static String expandYearToDate(String value) {
+//        return value.length() == 4 ? value + JANUARY_FIRST : value;
+//    }
```
Comment (joachimjorgensen) [inline search-commons/src/main/java/no/unit/nva/search2/common/constant/Functions.java:118]: Discard?

**Review by joachimjorgensen — APPROVED** [review]

**Pull request merged**